### ISSUE TITLE
Allow for a customized loss reduction when training with custom train_step in a MirroredStrategy

### ISSUE THREAD

**Issue by kretes** [keras-team/keras#15509]
**System information**.

TensorFlow version (you are using): 2.5.1
Are you willing to contribute it (Yes/No) : Yes

**Describe the feature and the current behavior/state**.

Training a model with custom train_step using a MirroredStrategy is reducing the loss values from replicas using a reduction method 'first'. This can be seen in https://github.com/tensorflow/tensorflow/blob/v2.5.1/tensorflow/python/keras/engine/training.py#L841 and we confirmed this by printing loss values in each replica and comparing this to the value that is being shown in the metrics print. We see that the printed value was taken from one of the replicas, it was not an aggregation.

Our model is build in a way that:
- it has multiple optimizers
- `train_step` and `test_step` are overriden and are doing forward pass + backprop
- The construction is done in the scope of a `MirroredStrategy`
- Training is done by model.fit()

We are following the guide in https://www.tensorflow.org/tutorials/distribute/custom_training and divide the loss on each replica by the global batch size.
This means to get the actual loss value we should sum up all the losses from replicas, however Keras is just doing a 'first' reduction

A suggested solution is to allow the sub-classes of the model to define their own reduction strategy, which is what I did by a simple modification in the Model code:
https://gist.github.com/kretes/62fac0bedff0d222516963e60a61ce05

**Will this change the current api? How?**

The API will remain the same, one new configuration option is available.

**Who will benefit from this feature?**

Developers training multi-gpu models with custom train step but with Model.fit() and not custom epochs loop

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): yes (if guided a bit)
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing): see the gist, It is probably required to do a change in the `make_test_unction` as well.

**Comment by SuryanarayanaY** [comment]
Hello, Thank you for reporting an issue.

We're currently in the process of migrating the new Keras 3 code base from keras-team/keras-core to keras-team/keras.
Consequently, This issue may not be relevant to the Keras 3 code base . After the migration is successfully completed, feel free to reopen this Issue at keras-team/keras if you believe it remains relevant to the Keras 3 code base. If instead this Issue is a bug or security issue in legacy tf.keras, you can instead report a new issue at [keras-team/tf-keras](https://github.com/keras-team/tf-keras/issues), which hosts the TensorFlow-only, legacy version of Keras.

To know more about Keras 3, please read https://keras.io/keras_core/announcement/
